cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/DefaultStore.java

```diff
@@ -283,51 +236,62 @@ public void setStop(final Id.Program id, final String pid, final long endTime, f
+                                  Id.Workflow.from(id.getApplication(), id.getId()), pid);
+        }
+        // todo: delete old history data
+      }
+    });
   }
 
-  private void recordCompletedWorkflow(final Id.Workflow id, String pid) {
-    final RunRecordMeta run = getRun(id, pid);
-    if (run == null) {
+  private void recordCompletedWorkflow(AppMetadataStore metaStore, WorkflowDataset workflowDataset,
+                                       Id.Workflow workflowId, String runId) {
+    RunRecordMeta runRecord = metaStore.getRun(workflowId, runId);
+    if (runRecord == null) {
       return;
     }
-    Id.Application app = id.getApplication();
-    ApplicationSpecification appSpec = getApplication(app);
-    if (appSpec == null || appSpec.getWorkflows() == null || appSpec.getWorkflows().get(id.getId()) == null) {
+    Id.Application app = workflowId.getApplication();
+    ApplicationSpecification appSpec = getApplicationSpec(metaStore, workflowId.getApplication());
+    if (appSpec == null || appSpec.getWorkflows() == null || appSpec.getWorkflows().get(workflowId.getId()) == null) {
       return;
     }
 
     boolean workFlowNodeFailed = false;
-    WorkflowSpecification workflowSpec = appSpec.getWorkflows().get(id.getId());
+    WorkflowSpecification workflowSpec = appSpec.getWorkflows().get(workflowId.getId());
     Map<String, WorkflowNode> nodeIdMap = workflowSpec.getNodeIdMap();
     final List<WorkflowDataset.ProgramRun> programRunsList = new ArrayList<>();
-    for (Map.Entry<String, String> entry : run.getProperties().entrySet()) {
+    for (Map.Entry<String, String> entry : runRecord.getProperties().entrySet()) {
       if (!("workflowToken".equals(entry.getKey()) || "runtimeArgs".equals(entry.getKey())
         || "workflowNodeState".equals(entry.getKey()))) {
         WorkflowActionNode workflowNode = (WorkflowActionNode) nodeIdMap.get(entry.getKey());
         ProgramType programType = ProgramType.valueOfSchedulableType(workflowNode.getProgram().getProgramType());
         Id.Program innerProgram = Id.Program.from(app.getNamespaceId(), app.getId(), programType, entry.getKey());
-        RunRecordMeta innerProgramRun = getRun(innerProgram, entry.getValue());
+        RunRecordMeta innerProgramRun = metaStore.getRun(innerProgram, entry.getValue());
         if (innerProgramRun != null && innerProgramRun.getStatus().equals(ProgramRunStatus.COMPLETED)) {
           Long stopTs = innerProgramRun.getStopTs();
           // since the program is completed, the stop ts cannot be null
-          Preconditions.checkState(stopTs != null, "Since the program has completed, expected its stop time to not " +
-            "be null. Program = %s, Workflow = %s, Run = %s, Stop Ts = %s", innerProgram, id, run, stopTs);
-          programRunsList.add(new WorkflowDataset.ProgramRun(
-            entry.getKey(), entry.getValue(), programType, stopTs - innerProgramRun.getStartTs()));
+          if (stopTs == null) {
```

Comment:
That's intended because this condition can only happen if the app gets delete right after the program stopped. This shouldn't be an exception, hence just log a warning.